hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergeHandle.java

```diff
@@ -452,6 +461,58 @@ protected HoodieRecord<T> updateFileName(HoodieRecord<T> record, Schema schema,
     return record.prependMetaFields(schema, targetSchema, metadataValues, prop);
   }
 
+  private void trackMetadataIndexStats(Option<HoodieKey> hoodieKeyOpt, Option<HoodieRecord> combinedRecordOpt, Option<HoodieRecord<T>> oldRecordOpt, boolean isDelete) {
```

Comment:
The `hoodieKeyOpt` is always present in the callers, can the method simply take in a `HoodieKey`

Reply:
I have made the option Nullable. We have fallback logic incase the key is null.
```
      Option<HoodieKey> hoodieKeyOpt = Option.ofNullable(newRecord.getKey());
```

Reply:
move all the index stats related changes into its corresbonding utility classes.

Reply:
> Also we are reusing a lot of instance variables from the handle class itself

what are they, just make the utility method static if you want and pass them around.

> so it would be better to place it in the handle class IMO

-1, I don't think so.